File: longest-substring-without-repeating-characters/Solution3.java
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

import java.util.HashMap;
import java.util.HashSet;

class Solution3 {

    // T: O(n), S: O(k) where k is the unique characters in the string
    public int lengthOfLongestSubstring(String s) {
        var seen = new HashSet<Character>();
        int left = 0, right = 0;
        int result = 0;
        while (right < s.length()) {
            if (seen.contains(s.charAt(right))) {
                seen.remove(s.charAt(left));
                left++;
            } else {
                seen.add(s.charAt(right));
                right++;
                result = Math.max(result, right - left);
            }
        }
        return result;
    }

    // T: O(n), S: O(k)
    public int lengthOfLongestSubstringOpt(String s) {
        var indices = new HashMap<Character, Integer>();
        int result = 0;

        for (int left = 0, right = 0; right < s.length(); right++) {
            char curr = s.charAt(right);

            // If the character is already in the current window,
            // move the start pointer to the right of its previous occurrence
            if (indices.containsKey(curr) && indices.get(curr) >= left) {
                left = indices.get(curr) + 1;
            }

            // Update the most recent index of the current character
            indices.put(curr, right);

            // Update the maximum length if the current window is longer
            result = Math.max(result, right - left + 1);
        }

        return result;
    }

    public static void main(String[] args) {
        var sol = new Solution3();
        var s1 = "abcabcbb";
        var s2 = "bbbbb";
        var s3 = "pwwkew";
        var s4 = "aab";
        System.out.println(sol.lengthOfLongestSubstring(s1)); // 3
        System.out.println(sol.lengthOfLongestSubstringOpt(s1)); // 3
        System.out.println(sol.lengthOfLongestSubstring(s2)); // 1
        System.out.println(sol.lengthOfLongestSubstringOpt(s2)); // 1
        System.out.println(sol.lengthOfLongestSubstring(s3)); // 3
        System.out.println(sol.lengthOfLongestSubstringOpt(s3)); // 3
        System.out.println(sol.lengthOfLongestSubstring(s4)); // 2
        System.out.println(sol.lengthOfLongestSubstringOpt(s4)); // 2
    }

}

File: longest-substring-without-repeating-characters/solution3.py
# https://leetcode.com/problems/longest-substring-without-repeating-characters/

class Solution3:

    # T: O(n), S: O(k) where k is the unique characters in the string
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen = set()
        left, right = 0, 0
        result = 0
        while right < len(s):
            if s[right] in seen:
                seen.remove(s[left])
                left += 1
            else:
                seen.add(s[right])
                right += 1
                result = max(result, right - left)

        return result

    # T: O(n), S: O(k
    def lengthOfLongestSubstringOpt(self, s: str) -> int:
        indices = dict()
        result = 0
        left = 0
        for right in range(len(s)):
            if s[right] in indices and indices[s[right]] >= left:
                left = indices[s[right]] + 1

            indices[s[right]] = right

            result = max(result, right - left + 1)

        return result


def main() -> None:
    sol = Solution3()
    s1 = "abcabcbb"
    s2 = "bbbbb"
    s3 = "pwwkew"
    s4 = "aab"
    print(sol.lengthOfLongestSubstring(s1))  # 3
    print(sol.lengthOfLongestSubstringOpt(s1))  # 3
    print(sol.lengthOfLongestSubstring(s2))  # 1
    print(sol.lengthOfLongestSubstringOpt(s2))  # 1
    print(sol.lengthOfLongestSubstring(s3))  # 3
    print(sol.lengthOfLongestSubstringOpt(s3))  # 3
    print(sol.lengthOfLongestSubstring(s4))  # 2
    print(sol.lengthOfLongestSubstringOpt(s4))  # 2


if __name__ == "__main__":
    main()
